app/lib/admin/backend.dart

```diff
@@ -400,28 +400,29 @@ class AdminBackend {
             'Last version detected. Use full package removal without the version qualifier.');
       }
 
-      bool updatePackage = false;
+      bool updateVersionReferences = false;
       if (package != null && package.latestVersion == version) {
         package.latestVersionKey = null;
         package.latestPublished = null;
-        updatePackage = true;
+        updateVersionReferences = true;
       }
       if (package != null && package.latestPrereleaseVersion == version) {
         package.latestPrereleaseVersionKey = null;
         package.latestPrereleasePublished = null;
-        updatePackage = true;
+        updateVersionReferences = true;
       }
       if (package != null && package.latestPreviewVersion == version) {
         package.latestPreviewVersionKey = null;
         package.latestPreviewPublished = null;
       }
-      if (updatePackage) {
+      if (updateVersionReferences) {
         package!.lastVersionPublished = null;
         versions.where((v) => v.version != version).forEach((v) => package
             .updateVersion(v, dartSdkVersion: currentDartSdk.semanticVersion));
-        package.updated = DateTime.now().toUtc();
-        tx.insert(package);
       }
+      package!.versionCount = package.versionCount! - 1;
```

Comment:
This doesn't seem like an idempotent operation... We can only decrease the version count if the same `tx` is deleting a `PackageVersion` entity.

If the request fails half way through and is tried, that might not be the case on the second try. Is that a fair analysis?

------

Unrelated should we printing in: https://github.com/dart-lang/pub-dev/blob/82861ec08e236f39f94bec0d71cf385a4b770479/app/lib/admin/backend.dart#L386

Reply:
Good catch!

> We can only decrease the version count if the same tx is deleting a PackageVersion entity.

It was the same tx, but we should only decrease when we also known there is a delete. Fixed that in the commit + added tests for it.